### ISSUE TITLE
Errors and GNU compatibility

### ISSUE THREAD

**Issue by tertsdiepraam** [uutils/coreutils#4000]
TL;DR: **I think we need a policy for error compatibility with GNU.** We can pick either:
- Errors should be identical to GNU.
- Errors can differ from GNU and we should aim to make them better than GNU.

<hr>

We advertise a goal of (almost) 100% GNU compatibility and I think we need to figure out whether that includes error messages. We currently have some parts of the code which try really hard to give the same errors and other parts which blatantly ignore the GNU messages. As a community, we need to decide what we want do to here.

**My personal opinion is that we should not make error messages identical to GNU**, because I think that this is an area where we could provide a significant improvement over GNU, by making prettier and more informative messages. The code also usually becomes simpler because we need to jump through fewer hoops to match GNU. However, this means that we need to patch the GNU test suite a lot where they test for specific messages.

> Note: I'm **not** advocating that we change _when_ messages pop up, only what the messages look like.

Questions to discuss:
1. Is an improvement in error messages worthwhile?
2. Are there scripts which rely on specific messages that we would break by introducing different messages?
3. Would you expect a drop-in replacement for coreutils to have identical error messages?
4. What is your personal preference on this?

Once we settle on something, I propose we add it to the documentation and the contributors guide.

Random selection of some relevant issues/PRs:
- https://github.com/uutils/coreutils/issues/2639
- https://github.com/uutils/coreutils/pull/3992
- https://github.com/uutils/coreutils/pull/3951
- https://github.com/uutils/coreutils/pull/3831
- https://github.com/uutils/coreutils/issues/818
- https://github.com/uutils/coreutils/issues/2336
- https://github.com/uutils/coreutils/issues/2527

**Comment by tertsdiepraam** [comment]
@sylvestre, @jfinkels what are your opinions on this?

**Comment by hbina** [comment]
What is GNU's policy on error messages? Do they expect people to rely on it to be consistent? Do people actually rely on it?

**Comment by tertsdiepraam** [comment]
Oh very good question. I can't find any policy in writing. Looking at their NEWS doc though. They seem to be okay with changing error messages even on minor version bumps (e.g. https://github.com/coreutils/coreutils/blob/4fd708810ce0e0d967c4c14e1ff2ff7b43440b58/NEWS#L1102), though it happens extremely rarely.

**Comment by sssemil** [comment]
Somewhat related - #4023, should there be any message for the 'no' choice?

**Comment by tertsdiepraam** [comment]
@sssemil That's a good example of where deviating from GNU could be nice. I think the extra message is not really necessary, but I like the `[y/N]`. What do  you think?

**Comment by sssemil** [comment]
I would remove the error message. The `[y/N]` is indeed nice :)

**Comment by sylvestre** [comment]
I agree with you. 

We should aim at:
* providing better error messages
* adjust the GNU testsuite (like we are here: https://github.com/uutils/coreutils/blob/main/util/build-gnu.sh#L157-L165 ) to make sure that the GNU test doesn't fail because of a difference of output

**Comment by jfinkels** [comment]
> We should aim at:
> 
> * providing better error messages
> * adjust the GNU testsuite (like we are here: https://github.com/uutils/coreutils/blob/main/util/build-gnu.sh#L157-L165 ) to make sure that the GNU test doesn't fail because of a difference of output

That makes sense to me. :+1:

**Comment by jfinkels** [comment]
related to #3572

**Comment by mon-jai** [comment]
Is this actually solved?
